src/operator/subgraph/build_subgraph.cc

```diff
@@ -360,6 +405,17 @@ void SelectSubgraphNodes(nnvm::Graph* g, SubgraphSelectorV2Ptr subgraph_selector
     // filter out unqualified pre-selected nodes
     std::vector<BiDirectedNode*> filtered_nodes = subgraph_selector->Filter(preselected_nodes);
 
+    const SubgraphPropertyPtr& subg_prop = g->GetAttr<SubgraphPropertyPtr>("subgraph_property");
+    if (subg_prop->HasAttr("ensure_CachedOp_input")
+        && subg_prop->GetAttr<bool>("ensure_CachedOp_input")) {
+      // check if subgraph has external input.
+      // if not, reject the first op (in top order) from the subgraph
+      // to make sure CachedOp gets external input.
+      if (filtered_nodes.size() > 0 && !HasInputEntries(*g, simple_nodes, filtered_nodes)) {
+        filtered_nodes.erase(filtered_nodes.begin());
```

Comment:
@waytrue17 are we handling the case when this leads to `filtered_nodes` containing only one node? we should ideally reject such a subgraph.

Reply:
Yes, this is handled by the filter function in our subgraph property.

Reply:
Can you double check? We don’t run filter function after removing first node from subgraph. In the case when preselected nodes contain exactly two ops, filter function does not do anything, right? And then if one node gets removed, do we discard the resulting subgraph containing a single node? Can you show me how that works?

Reply:
Thanks for bringing this up! I agreed that the filter function should've been called after removing first node, not before. I'll include the change in the next commit

Reply:
why are we calling `Filter` twice? Can we move `ensure_CachedOp_input` before the `Filter` call and check for Input Entries in `PreSelectSubgraphNodes`?

Reply:
Moved `ensure_CachedOp_input` inside `PreSelectSubgraphNodes`